Suggestion changes to the spec, (#71)
zarr-developers/zarr-specs: docs/protocol/core/v3.0.rst

This document was produced by the `Zarr core development team
<https://github.com/orgs/zarr-developers/teams/core-devs>`_.

> Goal of v3 spec and main difference from v2

Reviewer: ```suggestion
Goal of v3 spec and main difference with v2
```

===========================================

Zarr spec v2 was originally designed around local filesystem, but Zarr has
grown and is now regularly deployed on cloud / object storage. Those kind of
storage have characteristics, capabilities and usage patterns that can widely

> differ from the assumptions of spec v2. V3 is design to consider online
> stores in particular we want to achieve the following.

Reviewer: ```suggestion
differ from the assumptions of spec v2. V3 is designed to consider online
stores, in particular we want to achieve the following:
```

>  - No assumption that the underlying store have locking ability.
>  - Ability to do concurrent writes with the assumption that writes from clients will be consistent, but non atomic.

Reviewer: ```suggestion
 - No assumption that the underlying store has locking ability.
 - Ability to do concurrent writes with the assumption that writes from clients will be consistent, but not atomic.
```

> Unlike Zarr spec version 2, the spec V3 have mainly the following difference:
>   - V3 is a flat key-value store instead of an hierarchical store. Hierarchy is implied.
>   - V3 has an explicit root, while v2 roots and group could not be distinguished.

Reviewer: ```suggestion
Unlike Zarr spec v2, the spec v3 has mainly the following differences:
  - V3 is a flat key-value store instead of a hierarchical store. Hierarchy is implied.
  - V3 has an explicit root, while v2 roots and groups could not be distinguished.
```

  - Separation of the data and  metadata key space.
  - Explicit support for extensions.
  - chunk separator is ``/`` by default.

> This means that a store cannot be opened at an arbitrary point, and need to be
> opened at the root. User facing convenience functions could walk up a given

Reviewer: ```suggestion
This means that a store cannot be opened at an arbitrary point, but needs to be
opened at the root. User facing convenience functions could walk a given
```

This document was produced by the `Zarr core development team
<https://github.com/orgs/zarr-developers/teams/core-devs>`_.

Goal of v3 spec and main difference from v2
===========================================

Zarr spec v2 was originally designed around local filesystem, but Zarr has
grown and is now regularly deployed on cloud / object storage. Those kind of
storage have characteristics, capabilities and usage patterns that can widely
differ from the assumptions of spec v2. V3 is design to consider online
stores in particular we want to achieve the following.

 - No assumption that the underlying store have locking ability.
 - Ability to do concurrent writes with the assumption that writes from clients will be consistent, but non atomic.

Unlike Zarr spec version 2, the spec V3 have mainly the following difference:

>   - V3 is a flat key-value store instead of an hierarchical store. Hierarchy is implied. 

Reviewer 2: What are the implications of this on the local filesystem implementations? cc: @jni

Author: Not much:
 - you can't open a hierarchy at an arbitrary point, you have to open the root. (zarr_open, could walk up the tree until is finds the root, and return a subgroup, that's not an issue, but that's just user facing).
 - meta/ and data/ are in two subtrees, so the storage format is one layer deeper than v2

Also the chink separator is now `/` by default; which has a much bigger impact on the disk layout as now you have a deeper hierarchy, with less files per folder.

Reviewer 3: > Also the chink separator is now / by default; which has a much bigger impact on the disk layout as now you have a deeper hierarchy, with less files per folder.

Ah, that's the crux. This particular sentence makes it sound like the opposite is true, but they are talking about different things. Thanks for clarifying @Carreau! I'm not sure whether it makes sense to add a similar clarification with this bullet point — I'll leave that to people more familiar with this whole ecosystem. =)

* must not be a string composed only of period characters, e.g. "." or
  ".."

> TODO: likely want to reserver names that start with just a single ``.``

Reviewer 4: FWIW I don't think we need to reserve any names.

     - 2
     - big-endian
   * - ``>f4``
     - single precision float: sign bit, 8 bits exponent, 23 bits mantissa
     - 4
     - big-endian
   * - ``>f8``
     - double precision float: sign bit, 11 bits exponent, 52 bits mantissa
     - 8
     - big-endian
   * - ``r*`` (Optional)
     - raw bits,  use for extension type fallbacks
     - variable, given by ``*``, is not limited to be a multiple of 8.
     - N/A

Floating point types correspond to basic binary interchange formats as
defined by IEEE 754-2008.

Additionally to these base types, an implementation should also handle the
raw/opaque pass through type designated by the lowercase letter ``r`` followed
by the number of bits. For example, ``r3``, ``r11`` should be understood as
fallback types of respectively 3 and 11 bytes longs, and implementation should
ensure to pass or accept those types unchanged when using extension fallback.
For array whose total length is not a multiple of 8, zarr implementation can
expect expects a buffer which length is a multiple of 8bits, padded with
appropriate value at the end, and will in the same way return buffers padded to

> the next multiple of 8 bytes. 

Reviewer 5: This paragraph confuses bits and bytes a lot. First it says bits, then it says bytes, then bits again, then bytes. I think they should all say bits.

Author: bits everywhere; sorry, thanks.

   * - ``>i4``
     - signed integer
     - 4
     - big-endian
   * - ``>i8``
     - signed integer
     - 8
     - big-endian
   * - ``u1``
     - signed integer
     - 1
     - None
   * - ``<u2``
     - unsigned integer
     - 2
     - little-endian
   * - ``<u4``
     - unsigned integer
     - 4
     - little-endian
   * - ``<u8``
     - unsigned integer
     - 8
     - little-endian
   * - ``<f2``
     - half precision float: sign bit, 5 bits exponent, 10 bits mantissa
     - 2
     - little-endian
   * - ``<f4``
     - single precision float: sign bit, 8 bits exponent, 23 bits mantissa
     - 4
     - little-endian
   * - ``<f8``
     - double precision float: sign bit, 11 bits exponent, 52 bits mantissa
     - 8
     - little-endian
   * - ``>f2``
     - half precision float: sign bit, 5 bits exponent, 10 bits mantissa
     - 2
     - big-endian
   * - ``>f4``
     - single precision float: sign bit, 8 bits exponent, 23 bits mantissa
     - 4
     - big-endian
   * - ``>f8``
     - double precision float: sign bit, 11 bits exponent, 52 bits mantissa
     - 8
     - big-endian
   * - ``r*`` (Optional)
     - raw bits/bytes,  use for extension type fallbacks

>      - variable, given by ``*``, multiple of 8

Author: We'll discuss that today in the spec, my main questions was whether we want that to be the number of bits, or the number of bytes.

If it's number of Bytes, it can likely be any int.
The question is if we allow bits for types shorter the one byte (categorical or boolean), and define a behavior to pad/truncate.

>      - variable, given by ``*``, multiple of 8

Reviewer 6: @QuLogic, it would be nice to get your thoughts on this (and other bit related points). Would like to make sure this makes sense for architectures that you are thinking about 🙂

>      - variable, given by ``*``, multiple of 8

Reviewer 5: I think generally all systems in Fedora use 8-bit bytes, so there is no worry there. The key issue is order. I see there are definitions for endian order on multi-byte values, but I don't see anything about bit order (for a `packbits`-equivalent type, it looks like). If there's padding, is it between every entry, or just at the end? If the number of bits is > 8, is it first byte, then second, or the other way? Are the >8 bits in the lower or upper bits of the next byte? This is somewhat explained in examples, but not exactly explicit.

>      - variable, given by ``*``, multiple of 8

Author: Padding would just be at the end; not in between. I have no objections in changing. I think whether the order is lower first or lower last would be up to the extensions, or library using this which actually know the internal types. `r` is mostly meant as a pass through for when you can't fallback on an another of the type you can fallback on. Though I don't have that much experience with different endianness, so if you think this should be explicit I'll trust you.

     - big-endian
   * - ``>f8``
     - double precision float: sign bit, 11 bits exponent, 52 bits mantissa
     - 8
     - big-endian
   * - ``r*`` (Optional)
     - raw bits,  use for extension type fallbacks
     - variable, given by ``*``, is not limited to be a multiple of 8.
     - N/A

Floating point types correspond to basic binary interchange formats as
defined by IEEE 754-2008.

Additionally to these base types, an implementation should also handle the
raw/opaque pass through type designated by the lowercase letter ``r`` followed
by the number of bits. For example, ``r3``, ``r11`` should be understood as
fallback types of respectively 3 and 11 bits longs, and implementation should
ensure to pass or accept those types unchanged when using extension fallback.
For array whose total length is not a multiple of 8 bits, Zarr implementation can
expect a buffer which length is a multiple of 8 bits, padded with
appropriate value at the end, and will in the same way return buffers padded to
the next multiple of 8 bits. 

For example, here is a array of 5 elements of opaque type ``r3`` will be stored as

 -  ``AAABBBCC CDDDEEE_``, 3*5 = 15 bits, rounded to 2 bytes, with arbitrary

>    last bit.

Author: "bullet list ends w/o a blank line".

     - Numerical type
     - Size (no. bytes)
     - Byte order
   * - ``bool``
     - Boolean, with False encoded as ``\\x00`` and True encoded as ``\\x01``
     - 1
     - None
   * - ``i1``
     - signed integer
     - 1
     - None
   * - ``<i2``
     - signed integer
     - 2
     - little-endian
   * - ``<i4``
     - signed integer
     - 4
     - little-endian
   * - ``<i8``
     - signed integer
     - 8
     - little-endian
   * - ``>i2``
     - signed integer
     - 2
     - big-endian
   * - ``>i4``
     - signed integer
     - 4
     - big-endian
   * - ``>i8``
     - signed integer
     - 8
     - big-endian
   * - ``u1``

>      - signed integer

Reviewer: unsigned integer?

Author: probably my fault when I typed a vim macro. 

I gave you access to my fork if you want to make changes directly.

     - 1
     - None
   * - ``<u2``
     - unsigned integer
     - 2
     - little-endian
   * - ``<u4``
     - unsigned integer
     - 4
     - little-endian
   * - ``<u8``
     - unsigned integer
     - 8
     - little-endian
   * - ``<f2``
     - half precision float: sign bit, 5 bits exponent, 10 bits mantissa
     - 2
     - little-endian
   * - ``<f4``
     - single precision float: sign bit, 8 bits exponent, 23 bits mantissa
     - 4
     - little-endian
   * - ``<f8``
     - double precision float: sign bit, 11 bits exponent, 52 bits mantissa
     - 8
     - little-endian
   * - ``>f2``
     - half precision float: sign bit, 5 bits exponent, 10 bits mantissa
     - 2
     - big-endian
   * - ``>f4``
     - single precision float: sign bit, 8 bits exponent, 23 bits mantissa
     - 4
     - big-endian
   * - ``>f8``
     - double precision float: sign bit, 11 bits exponent, 52 bits mantissa
     - 8
     - big-endian
   * - ``r*`` (Optional)
     - raw bits,  use for extension type fallbacks
     - variable, given by ``*``, is limited to be a multiple of 8.

>      - N/A

Reviewer: I think this is too confusing to describe it as bits. The size column is in bytes, and we are saying it is limited to be a multiple of 8 bits anyway. I think we should just say "raw bytes".

Reviewer: Another reason is that all the other identifiers have a value in number of bytes, e.g. `i1` which is a 1-byte integer.

Author: Sure, but then we'll never be able to go lower or non multiple of 8.

The identifier for chunk with grid index (``i``, ``j``, ``k``, ...) is
formed by joining together ASCII string representations of each index
using a separator and prefixed with the string `'c'`. The default value for the separator is the period

> character ( by default ``/``), but this may be configured by providing a ``separator``

Reviewer: ```suggestion
using a separator and prefixed with the string `'c'`. The default value for the separator is the slash
character (by default ``/``), but this may be configured by providing a ``separator``
```

``metadata_encoding``

    A string containing the URI that has been used for encoding
    group and array metadata documents. If JSON encoding is used, this

>     value must be ``"https://purl.org/zarr/spec/metadata/json/3.0"``.

Reviewer 4: In general we want URIs to dereference to some documentation that defines the metadata encoding. But for JSON, I don't think we need to create a new URI, we could just point to an existing URI. E.g., we could use the JSON spec URI: `https://tools.ietf.org/html/rfc8259`.

Author: That is fine with me, but one of the reason we discuss to use URI instead of mimetype was the the same mimetype can encore metadata in a different way. So pointing to the JSON spec URI seem to me like it has the same issue.

You do would to not only point to the format of the file, but also it's structure; At least that was my understanding.

    {
        "zarr_format": "https://purl.org/zarr/spec/protocol/core/3.0",
        "metadata_encoding": "https://tools.ietf.org/html/rfc8259",

>         "extensions": [

Author: let's go with `https://purl.org/zarr/spec/protocol/core/3.0` + anchor to metadata section.

For example, the JSON document below defines an explicit group::

    {
        "attributes": {
            "spam": "ham",
            "eggs": 42,
        }
    }

.. note:: 

   Groups cannot have extensions attached to them as of spec v3.0 Allowing
   groups to have extensions would force any implementation to sequentially
   traverse the store hierarchy in order to check for extensions, which would
   defeat the purpose of a flat namespace and concurrent access.

   For the time being groups can only have attributes.

.. note::
   
   A group does not need a metadata document to exists, see implicit groups.

Chunks Metadata
---------------

Do we want to allow per-chunk metadata ? example checksum (concurrent read/writes).

Metadata encoding
-----------------

The entry point metadata document must be encoded as JSON. The array (``*.array``s) and
group metadata documents (``*.group``s) must be encoded as per the MIME type given in
the ``metadata_encoding`` field in the entry point metadata document
(described below).

@@TODO more explanation needed?
Todo questions : How frequent are non-JSON metadata documents ? and how store specific are those. 

> If we allow arbitrary mimetypes, does that precude the sections on arrays and group metadata to be normative ? 

Author: Maybe use URI. 
Woudl also allow to encode institution.

**Delete a group or array**

    To delete an array it is necessary to 
      - delete the metadata document for the array, (meta/P.array)
      - delete all keys which prefix have path pointing to this to this array.  (data/P/\*)

    To delete a implicit group. 
      - delete all arrays under this group
      - it should be sufficient to delete all the keys starting with prefix meta/P/ and data/P/

    To delete an explicit group. 
      - delete all arrays under this group, 
      - delete all keys with meta/P/ prefix, meta/P/groups all keys with /data/P prefix, 

    Note that store implementation may decide to reify implicit groups and thus protocol implementation should attempt to delete the .meta/P/.group file if they really wish to delte an empty implicit group. 

    Store implementation are also allowed to delete any implicit parent of a deleted implicit groups, so a protocol implementation should make sure to reify a parent group if they need to keep it.

**Determine if a node exists**

    To determine if a node exists at path `P`, perform `exist()` 
    exists is tricky, you can query `meta/P.*`, as P might be implicit, 

>     so you need to query both `meta/P.*` and `meta/P/*`.

Reviewer 4: In [zarrita](https://github.com/alimanfoo/zarrita/blob/master/zarrita.py#L323) I try a sequence of operations to determine if a node exists at hierarchy path `P`:

* ``get("meta/root{P}.array")`` - if returns a value, found array 
* ``get("meta/root{P}.group")`` - if returns a value, found explicit group
* ``list_dir("meta/root{P}/")`` - if results, found an implicit group

If all fail, no node exists at path `P`.
